Packet++/header/DoIpLayer.h

```diff
@@ -0,0 +1,1806 @@
+		static constexpr size_t MAX_DATA_SIZE_OFFSET = FIXED_LEN;
+		static constexpr size_t MAX_DATA_SIZE_LEN = 4;
+		static constexpr size_t OPT_LEN = FIXED_LEN + MAX_DATA_SIZE_LEN;
+	};
+
+	//~~~~~~~~~~~~~~~|
+	// DiagnosticBase|
+	//~~~~~~~~~~~~~~~|
+
+	/// @class DiagnosticBase
+	/// @brief Represents a DoIP Diagnostic Message sent between tester and ECU.
+	/// This message includes source and target addresses and carries diagnostic service data.
+	class DiagnosticBase : public DoIpLayer
+	{
+	public:
+		/// @brief Constructs the DiagnosticMessage from raw packet data.
+		/// @param[in] data Pointer to the raw payload data.
+		/// @param[in] dataLen Length of the data buffer.
+		/// @param[in] prevLayer Pointer to the previous protocol layer.
+		/// @param[in] packet Pointer to the parent packet.
+		DiagnosticBase(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet);
+
+		/// @brief default c'tor.
+		DiagnosticBase() {};
+
+		/// @struct  common_diagnostic_header
+		/// Common first diagnostic data in DoIP
+		/// messages (diagnostic/diagnosticAck/diagnosticNack).
+		/// common_diagnostic_header message structure (extends DoIP header).
+		struct common_diagnostic_header : doiphdr
+		{
+			/// @brief Diagnostic source address
+			uint16_t sourceAddress;
+
+			/// @brief Diagnostic target address
+			uint16_t targetAddress;
+		};
+
+		/// @brief Gets the source logical address of the message.
+		/// @return 16-bit address of the source ECU.
+		uint16_t getSourceAddress() const;
+
+		/// @brief Gets the target logical address of the message.
+		/// @return 16-bit address of the destination ECU.
+		uint16_t getTargetAddress() const;
+
+		/// @brief Sets the source logical address.
+		/// @param[in] addr New 16-bit source address.
+		void setSourceAddress(uint16_t addr);
+
+		/// @brief Sets the target logical address.
+		/// @param[in] addr New 16-bit target address.
+		void setTargetAddress(uint16_t addr);
+
+		/// @brief Returns a human-readable summary of the message content.
+		/// @return A string summarizing the diagnostic message.
+		virtual std::string getSummary() const = 0;
+
+		/// Get a pointer to diagnostic message data
+		/// Notice this points directly to the data, so any change will modify
+		/// the actual packet data
+		/// @return A pointer to a common_diagnostic_header structure containing the data
+		common_diagnostic_header* getCommonDiagnosticHeader() const
+		{
+			return reinterpret_cast<common_diagnostic_header*>(m_Data);
+		}
+
+	private:
+		static constexpr size_t FIX_LEN = DOIP_SOURCE_ADDRESS_LEN + DOIP_TARGET_ADDRESS_LEN;
+	};
+
+	//~~~~~~~~~~~~~~~~~~|
+	// DiagnosticMessage|
+	//~~~~~~~~~~~~~~~~~~|
+
+	/// @class DiagnosticMessage
+	/// @brief Represents a DoIP Diagnostic message sent between tester and ECU.
+	///
+	/// This message includes source and target addresses and carries diagnostic service data.
+	class DiagnosticMessage : public DiagnosticBase
```

Comment:
This class should override `getHeaderLen()` because its length is not `m_DataLen`, but the minimum between `sizepf(doiphdr) + 2 * sizeof(uint16_t)` and `m_DataLen`